addons/stock/models/stock_storage_category.py

```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import _, api, fields, models
+
+
+class StorageCategory(models.Model):
+    _name = 'stock.storage.category'
+    _description = "Storage Category"
+    _order = "name"
+
+    name = fields.Char('Storage Category', required=True)
+    max_weight = fields.Float('Max Weight', digits='Stock Weight')
+    capacity_ids = fields.One2many('stock.storage.category.capacity', 'storage_category_id')
+    product_capacity_ids = fields.One2many('stock.storage.category.capacity', compute="_compute_storage_capacity_ids", inverse="_set_storage_capacity_ids")
+    package_capacity_ids = fields.One2many('stock.storage.category.capacity', compute="_compute_storage_capacity_ids", inverse="_set_storage_capacity_ids")
+    allow_new_product = fields.Selection([
+        ('empty', 'If the location is empty'),
+        ('same', 'If all products are same'),
+        ('mixed', 'Allow mixed products')], default='mixed', required=True)
+    company_id = fields.Many2one('res.company', 'Company')
+
+    _sql_constraints = [
+        ('positive_max_weight', 'CHECK(max_weight >= 0)', 'Max weight should be a positive number.'),
+    ]
+
+    @api.depends('capacity_ids')
+    def _compute_storage_capacity_ids(self):
+        for storage_category in self:
+            storage_category.product_capacity_ids = storage_category.capacity_ids.filtered(lambda c: c.product_id)
+            storage_category.package_capacity_ids = storage_category.capacity_ids.filtered(lambda c: c.package_type_id)
+
+    def _set_storage_capacity_ids(self):
+        for storage_category in self:
+            storage_category.capacity_ids = storage_category.product_capacity_ids | storage_category.package_capacity_ids
+
+    @api.returns('self', lambda value: value.id)
+    def copy(self, default=None):
+        default = dict(default or {})
+        default.update(name=_("%s (copy)") % self.name)
+        res = super().copy(default)
+        for capacity in self.capacity_ids:
+            capacity.copy({'storage_category_id': res.id})
+        return res
+
+
+class StorageCategoryProductCapacity(models.Model):
```

Comment:
not sure it answers your question, but we intend in another development to link a packaging with a package type. this way when you order packagings (something we'll add as well). Assume you set a 6-pack (packaging) of a product  as Small box (package type). Then when receiving, it will automatically use the rule for this package type (not necessary to  do a put in pack)